File: iosApp/iosApp/resources/text/FormattedText.swift
import Foundation
import Shared

extension StringRef {
    func asString() -> String {
        return StringRefResolver.resolve(ref: self)
    }
}

extension StringRefWithArgs {
    func asStringWithArgs(_ firstArg: String, _ otherArgs: String...) -> String {
        let allArgs = [firstArg] + otherArgs
        return StringRefResolver.resolve(ref: self, args: allArgs)
    }
    
    func asStringWithArg(_ arg: String) -> String {
        return StringRefResolver.resolve(ref: self, arg: arg)
    }
}

extension FormattedText.Ref {
    func with(_ firstArg: String, _ otherArgs: String...) -> String {
        let allArgs = [firstArg] + otherArgs
        return StringRefResolver.resolve(ref: self.ref, args: allArgs)
    }
    
    func resolveString() -> String {
        return with("")
    }
}

extension FormattedText.WithString {
    func resolveString() -> String {
        return self.ref.asStringWithArg(self.arg)
    }
}

extension FormattedText.WithRef {
    func resolveString() -> String {
        return self.ref.asStringWithArg(self.arg.asString())
    }
}

extension FormattedText.SimpleText {
    func resolveString() -> String {
        return self.textRef.asString()
    }
}

extension FormattedText {
    func asString() -> String {
        if let t = self as? FormattedText.Ref { return t.resolveString() }
        if let t = self as? FormattedText.WithString { return t.resolveString() }
        if let t = self as? FormattedText.WithRef { return t.resolveString() }
        if let t = self as? FormattedText.SimpleText { return t.resolveString() }
        return self.ref.ref.simpleName()
    }
}

File: composeApp/src/commonMain/kotlin/com/exploramus/app/resources/FormattedText.kt
package com.exploramus.app.resources

import androidx.compose.runtime.Composable
import com.exploramus.shared.resources.FormattedText
import com.exploramus.shared.resources.StringRef
import com.exploramus.shared.resources.StringRefWithArgs

@Composable
fun StringRef.asString(): String = StringRefResolver.resolve(this)

@Composable
fun StringRefWithArgs.asStringWithArgs(firstArg: String, vararg otherArgs: String): String =
    StringRefResolver.resolve(this, firstArg, *otherArgs)

@Composable
fun StringRefWithArgs.asStringWithArg(arg: String): String =
    StringRefResolver.resolve(this, arg)

@Composable
fun FormattedText.asString(): String {
    return when (this) {
        is FormattedText.Ref -> this.ref.ref.asString()
        is FormattedText.WithString -> this.ref.asStringWithArg(this.arg)
        is FormattedText.WithRef -> this.ref.asStringWithArg(this.arg.asString())
        is FormattedText.SimpleText -> this.textRef.asString()
    }
}

@Composable
fun FormattedText.Ref.with(firstArg: String, vararg otherArgs: String): String {
    return ref.asStringWithArgs(firstArg, *otherArgs)
}
